src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,1421 @@
+
+            reason = null;
+
+            if (settings.Recurrence.Range.NumberOfOccurrences == null)
+            {
+                reason = RequiredParameter;
+
+                return false;
+            }
+
+            if (settings.Recurrence.Range.NumberOfOccurrences < 1)
+            {
+                reason = OutOfRange;
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool TryValidateRecurrenceTimeZone(TimeWindowFilterSettings settings, out string paramName, out string reason)
+        {
+            paramName = $"{nameof(settings.Recurrence)}.{nameof(settings.Recurrence.Range)}.{nameof(settings.Recurrence.Range.RecurrenceTimeZone)}";
+
+            reason = null;
+
+            if (settings.Recurrence.Range.RecurrenceTimeZone != null && !TryParseTimeZone(settings.Recurrence.Range.RecurrenceTimeZone, out _))
+            {
+                reason = UnrecognizableValue;
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool TryParseTimeZone(string timeZoneStr, out TimeSpan timeZoneOffset)
+        {
+            timeZoneOffset = TimeSpan.Zero;
+
+            if (timeZoneStr == null)
+            {
+                return false;
+            }
+
+            if (!timeZoneStr.StartsWith("UTC+") && !timeZoneStr.StartsWith("UTC-"))
+            {
+                return false;
+            }
+
+            if (!TimeSpan.TryParseExact(timeZoneStr.Substring(4), @"hh\:mm", null, out timeZoneOffset))
+            {
+                return false;
+            }
+
+            if (timeZoneStr[3] == '-')
+            {
+                timeZoneOffset = -timeZoneOffset;
+            }
+
+            return true;
+        }
+
+        private static TimeSpan GetRecurrenceTimeZone(TimeWindowFilterSettings settings)
+        {
+            if (!TryParseTimeZone(settings.Recurrence.Range.RecurrenceTimeZone, out TimeSpan timeZoneOffset))
+            {
+                timeZoneOffset = settings.Start.Value.Offset;
+            }
+
+            return timeZoneOffset;
+        }
+
+        /// <summary>
+        /// Check whether the duration is shorter than the minimum gap between recurrence of days of week.
+        /// </summary>
+        /// <param name="duration">The time span of the duration.</param>
+        /// <param name="interval">The recurrence interval.</param>
+        /// <param name="daysOfWeek">The days of the week when the recurrence will occur.</param>
+        /// <param name="firstDayOfWeek">The first day of the week.</param>
+        /// <returns>True if the duration is compliant with days of week, false otherwise.</returns>
+        private static bool IsDurationCompliantWithDaysOfWeek(TimeSpan duration, int interval, IEnumerable<string> daysOfWeek, string firstDayOfWeek)
+        {
+            if (daysOfWeek == null || !daysOfWeek.Any())
```

Comment:
The check about `daysOfWeek` is also repeated. We already have this check in `TryValidateDaysOfWeek` before.

Reply:
Removed.